src/hunspell/suggestmgr.cxx

```diff
@@ -1724,79 +1779,55 @@ std::string SuggestMgr::suggest_gen(const std::vector<std::string>& desc, const
   return result2;
 }
 
-// generate an n-gram score comparing s1 and s2
+// generate an n-gram score comparing s1 and s2, UTF16 version
 int SuggestMgr::ngram(int n,
-                      const std::string& s1,
-                      const std::string& s2,
+                      const std::vector<w_char>& su1,
+                      const std::vector<w_char>& su2,
                       int opt) {
   int nscore = 0;
   int ns;
   int l1;
   int l2;
   int test = 0;
 
-  if (utf8) {
-    std::vector<w_char> su1;
-    std::vector<w_char> su2;
-    l1 = u8_u16(su1, s1);
-    l2 = u8_u16(su2, s2);
-    if ((l2 <= 0) || (l1 == -1))
-      return 0;
-    // lowering dictionary word
-    if (opt & NGRAM_LOWERING)
-      mkallsmall_utf(su2, langnum);
-    for (int j = 1; j <= n; j++) {
-      ns = 0;
-      for (int i = 0; i <= (l1 - j); i++) {
-        int k = 0;
-        for (int l = 0; l <= (l2 - j); l++) {
-          for (k = 0; k < j; k++) {
-            w_char& c1 = su1[i + k];
-            w_char& c2 = su2[l + k];
-            if ((c1.l != c2.l) || (c1.h != c2.h))
-              break;
-          }
-          if (k == j) {
-            ns++;
+  l1 = su1.size();
+  l2 = su2.size();
+  if (l1 == 0 || l2 == 0)
```

Comment:
There is a small change in behavior here. The case where su1 is empty (l1 == 0) should continue and it should not return.
Additionally u8_u16 can return less than zero which means there were errors in the decoding of UTF8. Since we moved the conversion out of this method we should take care of that outside. Simply we wont call `ngram` if UTF decoding error occurred.